File: app/src/main/java/br/com/alura/leilao/model/Leilao.java
package br.com.alura.leilao.model;

import java.io.Serializable;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import br.com.alura.leilao.exception.QuantidadeMaximaDeLancesException;
import br.com.alura.leilao.exception.UsuarioDeuLancesSeguidosException;
import br.com.alura.leilao.exception.ValorMenorQueOAnteriorException;

public class Leilao implements Serializable {

	private final String descricao;
	private final List<Lance> lances;

	public Leilao(String descricao) {
		this.descricao = descricao;
		this.lances = new ArrayList<>();
		this.lances.add(new Lance(new Usuario(""), 0.0));
	}

	public void propoe(Lance lance) {
		if(lances.get(0).getValor() == 0.0) {
			lances.set(0, lance);
		} else {
			valida(lance);
			lances.add(0, lance);
		}
	}

	private void valida(Lance lance) {
		validaLanceSemRepeticaoDoUsuario(lance);
		validaQuantidadeMaximaDeLancesDeUmMesmoUsuario(lance);
		validaValorMaiorQueAnterior(lance);
	}

	private void validaLanceSemRepeticaoDoUsuario(Lance lance) {
		if(lance.getUsuario().equals(lances.get(0).getUsuario())) {
			throw new UsuarioDeuLancesSeguidosException();
		}
	}

	private void validaQuantidadeMaximaDeLancesDeUmMesmoUsuario(Lance lance) {
		int contador = 0;
		for(Lance l : lances) {
			if(lance.getUsuario().equals(l.getUsuario())) {
				contador++;
			}
		}
		if(contador == 5) {
			throw new QuantidadeMaximaDeLancesException();
		}
	}

	private void validaValorMaiorQueAnterior(Lance lance) {
		if(lance.getValor() < lances.get(0).getValor())
			throw new ValorMenorQueOAnteriorException();
	}

	public String getDescricao() {
		return descricao;
	}

	public List<Lance> getLances() {
		return lances;
	}

	public Lance getMaiorLance() {
		if(lances.isEmpty()) {
			return null;
		}
		return lances.get(0);
	}

	public String getMaiorLanceFormatado(){
		return getMaiorLance().getValorFormatado();
	}

	public Lance getMenorLance() {
		if(lances.isEmpty()) {
			return null;
		}
		return lances.get(lances.size() - 1);
	}

	public String getMenorLanceFormatado(){
		return getMenorLance().getValorFormatado();
	}

	public List<Lance> getTresMaioresLances() {
		if(lances.isEmpty()) {
			return lances;
		}
		if(lances.size() < 3) {
			return lances.subList(0, lances.size());
		}
		return lances.subList(0, 3);
	}
}

File: app/src/main/java/br/com/alura/leilao/model/Leilao.kt
package br.com.alura.leilao.model

import br.com.alura.leilao.exception.QuantidadeMaximaDeLancesException
import br.com.alura.leilao.exception.UsuarioDeuLancesSeguidosException
import br.com.alura.leilao.exception.ValorMenorQueOAnteriorException
import java.io.Serializable
import kotlin.jvm.Throws

data class Leilao(
    val id: Long = 0L,
    val descricao: String,
    val lances: MutableList<Lance> = mutableListOf(),
) : Serializable {

    init {
        if (lances.isEmpty()) {
            lances.add(Lance(Usuario(nome = ""), 0.0))
        }
    }

    val maiorLanceFormatado: String
        get() {
            return getMaiorLance().valorFormatado
        }
    val menorLanceFormatado: String
        get() {
            return getMenorLance().valorFormatado
        }
    val tresMaioresLances: List<Lance>
        get() {
            return when {
                lances.isEmpty() -> lances
                lances.size < 3 -> lances.subList(0, lances.size)
                else -> lances.subList(0, 3)
            }
        }

    fun propoe(lance: Lance) {
        if (lances[0].valor == 0.0) {
            lances[0] = lance
        } else {
            valida(lance)
            lances.add(0, lance)
        }
    }

    private fun valida(lance: Lance) {
        validaLanceSemRepeticaoDoUsuario(lance)
        validaQuantidadeMaximaDeLancesDeUmMesmoUsuario(lance)
        validaValorMaiorQueAnterior(lance)
    }

    private fun validaLanceSemRepeticaoDoUsuario(lance: Lance) {
        if (lance.usuario == lances[0].usuario) {
            throw UsuarioDeuLancesSeguidosException()
        }
    }

    private fun validaQuantidadeMaximaDeLancesDeUmMesmoUsuario(lance: Lance) {
        val contador = lances.count { it.usuario == lance.usuario }
        if (contador == 5) {
            throw QuantidadeMaximaDeLancesException()
        }
    }

    private fun validaValorMaiorQueAnterior(lance: Lance) {
        if (lance.valor < lances[0].valor) {
            throw ValorMenorQueOAnteriorException()
        }
    }

    private fun getMaiorLance(): Lance {
        return lances[0]
    }

    private fun getMenorLance(): Lance {
        return lances[lances.size - 1]
    }


}